prow/plugins/project/project.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package project implements the `/project` command which allows members of the project
+// maintainers team to specify a project to be applied to an Issue or PR.
+package project
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	pluginName = "project"
+)
+
+var (
+	projectRegex       = regexp.MustCompile(`(?m)^/project\s+(.+?)\s+(.+?)\s*$`)
+	notATeamMemberMsg  = "You must be a member of the [%s/%s](https://github.com/orgs/%s/teams/%s/members) github team to set the project and column."
+	invalidProject     = "The provided project is not valid for this organization. Projects in Kubernetes orgs and repositories: [%s]\n\n"
+	invalidColumn      = "A column is not provided or it's not valid for the project %s. The issue/PR is added to column %s by default.\n\n"
+	invalidNumArgs     = "Please provide 2 or 3 arguments. Example usage: /project 0.5.0, /project 0.5.0 3999801, /project clear 0.4.0"
+	projectTeamMsg     = "The project maintainers team is the Github team %q with ID: %d."
+	clearKeyword       = "clear"
+	projectNameToIDMap map[string]int
```

Comment:
Let's not use a global here, but we would need a project client similar to that which the other plugins have like ownersClient or slack client. Alternatively if we don't think it is _necessary_ to have an in-memory cache in this plugin for projects, we can just forget about it and offload the caching to the `ghproxy`. Sorry, I know I brought this up in the first place, but removing this might be the best approach for the first impl

Reply:
@stevekuznetsov I see. I think it might be overkill to have a client in this case. Could you point me to an example of how to offload the caching to ghproxy?

Reply:
Using `ghproxy` will be invisible to this code -- we just pass the proxy as the URL to GitHub when we start up the binary. So, to use it we can just ignore the fact in the code and keep things simple here.